### PR TITLE
Add automated transformer integration test

### LINKED ISSUE #223: Automated integration test for transformers
### Problem Description

To validate transformers end-to-end, there are certain behaviors that we expect from all transformers. We can automate these checks, so that they run for every transformer by default.

### Expected behavior

Automatically run end-to-end tests for each transformer, using the generated datasets that already exist for performance testing. These integration tests should validate that

1. The Transformer is correct on its own
    - The transformed data is the expected data type(s)
    - The reverse transformed data is the expected data type
    - The reverse transformed data is the same as the original data (for the transformers for which composition is identity)
    - The Transformer can be used by the HyperTransformer
   
2. If we specify the transformer in the hypertransformer, the hypertransformer can go from the specified input type to a numerical output, and back.

### PR DESCRIPTION
**Pull request by katxiao**
Add automated transformer integration test

Resolves #223

### CODE REVIEW

File: tests/integration/test_transformers.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                test_cases.append((t, ds, g))
+
+    return test_cases
+
+
+test_cases = get_transformer_test_cases()
+
+
+@pytest.mark.parametrize('transformer,data_size,dataset_generator', test_cases)
+def test_transformer(transformer, data_size, dataset_generator):
```
Comment (csala) [inline tests/integration/test_transformers.py:69]: I added a comment about this directly in the issue, so we can discuss there in a more organized way.

File: tests/integration/test_transformers.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    for g in BaseDatasetGenerator.__subclasses__():
+        generators[g.DATA_TYPE] = g.__subclasses__()
+
+    return generators
+
+
+def _find_dataset_generators(data_type, generators):
+    """Find the dataset generators for the given data_type."""
+    if data_type is None or data_type not in generators:
```
Comment (amontanez24) [inline tests/integration/test_transformers.py:41]: could we do something like 
```python3
return generators.get(data_type, [])
```

File: tests/integration/test_transformers.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    if isinstance(out, pd.DataFrame):
+        reversed_data = out[TEST_COL]
+    if isinstance(out, pd.DatetimeIndex):
+        reversed_data = out.to_series(index=pd.RangeIndex(start=0, stop=DATA_SIZE, step=1))
+    if isinstance(reversed_data, np.ndarray):
+        out = pd.Series(reversed_data)
+    if pd.api.types.is_datetime64_any_dtype(reversed_data):
+        reversed_data = reversed_data.round('us')
+
+    _validate_composition(transformer, reversed_data, input_data)
```
Comment (amontanez24) [inline tests/integration/test_transformers.py:121]: If I'm reading this correctly, all these checks only matter if `transformer.is_composition_identity() == True`. Should we move this stuff to after that check inside of `_validate_composition`? Or will it be used for other checks later?

**Review by amontanez24 — COMMENTED** [review]
Looking good, just some small comments

File: tests/integration/test_transformers.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+TEST_COL = 'test_col'
+
+# Mapping of rdt data type to dtype
+DATA_TYPE_TO_DTYPES = {
+    'boolean': ['b', 'O'],
+    'categorical': ['O', 'i', 'f'],
+    'datetime': ['M'],
+    'numerical': ['f', 'i'],
+    'integer': ['i'],
+    'float': ['f', 'i'],
```
Comment (katxiao) [inline tests/integration/test_transformers.py:19]: Right now, `OneHotEncodingTransformer` sometimes returns ints, but has an output type of `float`. I included `i` in this list to get the tests to pass. Should we update the output type of the transformer to `numerical`?

File: tests/integration/test_transformers.py (hunk 3 above)
Comment (katxiao) [reply tests/integration/test_transformers.py]: Yes we can move it into `_validate_composition` 👍

File: tests/integration/test_transformers.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    assert input_data_type is not None
+
+    dataset_generators = _find_dataset_generators(input_data_type, generators)
+    assert len(dataset_generators) > 0
+
+    for dg in dataset_generators:
+        with subtests.test(msg="test_transformer_with_dataset", generator=dg):
+            input_data = pd.Series(dg.generate(DATA_SIZE))
+
+            data = pd.DataFrame({TEST_COL: input_data})
```
Comment (csala) [inline tests/integration/test_transformers.py:186]: I think we could replace `input_data` with `dg.generate(DATA_SIZE)` and skip the line above.

File: tests/integration/test_transformers.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    Test the transformer end-to-end with at least one generated dataset. Test
+    both the transformer by itself, and by running in the hypertransformer.
+
+    Args:
+        transformer (rdt.transformers.BaseTransformer):
+            The transformer to test.
+    """
+    input_data_type = transformer.get_input_type()
+    assert input_data_type is not None
```
Comment (csala) [inline tests/integration/test_transformers.py:177]: I think we can skip this validation, and in the `_find_dataset_generators` function below return all the dataset generators that exist.

File: tests/integration/test_transformers.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    # Reverse transform
+    out = transformer.reverse_transform(transformed)
+    _validate_reverse_transformed_data(transformer, out, input_data.dtypes[TEST_COL])
+
+    _validate_composition(transformer, out, input_data)
+
+
+def _validate_hypertransformer_transformed_data(transformed_data):
+    """Check that the transformed data is of type float."""
+    for dt in transformed_data.dtypes:
```
Comment (csala) [inline tests/integration/test_transformers.py:126]: nit: `dtype`

File: tests/integration/test_transformers.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        assert dt.kind in DATA_TYPE_TO_DTYPES['numerical']
```
Comment (csala) [inline tests/integration/test_transformers.py:127]: I think we should also validate that there are no null values

File: tests/integration/test_transformers.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    This is only applicable if the transformer has the composition
+    identity property.
+    """
+    if not transformer.is_composition_identity():
+        return
+
+    if isinstance(reversed_data, pd.DataFrame):
+        reversed_data = reversed_data[TEST_COL]
+    if isinstance(reversed_data, pd.DatetimeIndex):
```
Comment (csala) [inline tests/integration/test_transformers.py:79]: I think these should be `elif`

**Review by csala — CHANGES_REQUESTED** [review]
I added a few comments, but in general this is looking great @katxiao !
I think that a few other changes will be required after PR #240 is merged too.

File: tests/integration/test_transformers.py (hunk 9 above)
Comment (katxiao) [reply tests/integration/test_transformers.py:79]: Multiple of these conditions could apply - for example, if the reversed_data is a DataFrame and a datetime. I could `elif` the two `isinstance` checks.

File: tests/integration/test_transformers.py (hunk 4 above)
Comment (katxiao) [reply tests/integration/test_transformers.py:19]: Opened https://github.com/sdv-dev/RDT/issues/268

File: tests/integration/test_transformers.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    """Build a map of data type to data generator.
+
+    Output:
+        dict:
+            A mapping of data type (str) to a list of data
+            generators (rdt.tests.datasets.BaseDatasetGenerator).
+    """
+    generators = {}
+
+    for g in BaseDatasetGenerator.__subclasses__():
```
[the remark this answers is not in the document]
Comment (katxiao) [reply tests/integration/test_transformers.py:33]: I see. Right now we have subclasses like `NumericalGenerator` (that other numerical generators subclass), which don't implement the `generate` method. I agree that we should allow for flexibility for subclasses to have grand-children. I think I will just have to catch the `NotImplementedError` for the case where `generate` is not implemented.

**Review by csala — APPROVED** [review]
Let's go!

**Review by csala — APPROVED** [review]
Re-approving this, and also re-requesting review from @amontanez24 
@amontanez24 @katxiao This and #240 will clash, because the HyperTransformer API changes a bit. I suggest we merge #240 first, then update this one, and then merge

File: tests/datasets/base.py [hunk 11]
```diff
@@ -20,3 +20,20 @@ def generate(num_rows):
             numpy.ndarray of size ``num_rows``
         """
         raise NotImplementedError()
+
+    @classmethod
+    def get_subclasses(cls):
```
Comment (amontanez24) [inline tests/datasets/base.py:25]: I notice we do this in a few places. Not really necessary for this PR, but maybe we should make this a utility function somewhere that can take any class and get all the subclasses

**Review by amontanez24 — APPROVED** [review]
LGTM!

File: tests/datasets/base.py (hunk 11 above)
Comment (katxiao) [reply tests/datasets/base.py:25]: Yeah that's what I had originally (put in a utils file), but I think @csala preferred it to be on the transformer. We can revisit this if it comes up again.

**Pull request merged**